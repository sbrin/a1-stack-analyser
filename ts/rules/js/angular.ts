import { register } from '../../register.js';

register({
  tech: 'angular',
  name: 'Angular',
  type: 'language',
  files: ['angular.json'],
  dependencies: [
    { type: 'npm', name: 'angular' },
    { type: 'npm', name: /^@angular\//, example: '@angular/forms' },
  ],
});
