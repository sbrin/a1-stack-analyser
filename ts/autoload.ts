import { loadAllRules } from './loader.js';
import './rules/index.js';

loadAllRules();
