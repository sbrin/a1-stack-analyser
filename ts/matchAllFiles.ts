import path from 'node:path';

import { rulesExtensions, rulesTechs } from './loader.js';

import type { AllowedKeys, ProviderFile } from './index.js';

export function matchAllFiles(
  files: ProviderFile[],
  basePath: string
): Map<AllowedKeys, string[]> {
  const matched = new Map<AllowedKeys, string[]>();

  // Match files
  for (const rule of rulesTechs) {
    const res = rule(files);
    if (!res) {
      continue;
    }

    matched.set(res[0].tech, [`matched file: ${res[1].replace(basePath, '')}`]);
  }

  // Match extensions
  const exts = new Set<string>();
  for (const file of files) {
    exts.add(path.extname(file.name));
  }
  for (const rule of rulesExtensions) {
    const res = rule(exts);
    if (!res) {
      continue;
    }
    if (matched.has(res[0].tech)) {
      continue;
    }

    matched.set(res[0].tech, [`matched extension: ${res[1]}`]);
  }

  return matched;
}
