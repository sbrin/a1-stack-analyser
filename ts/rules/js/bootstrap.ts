import { register } from '../../register.js';

register({
  tech: 'bootstrap',
  name: 'Bootstrap',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'bootstrap' }],
});
