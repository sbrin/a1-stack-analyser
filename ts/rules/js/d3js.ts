import { register } from '../../register.js';

register({
  tech: 'd3js',
  name: 'D3.js',
  type: 'tool',
  dependencies: [
    { type: 'npm', name: 'd3' },
    { type: 'npm', name: 'd3-shape' },
    { type: 'npm', name: '@types/d3' },
    { type: 'npm', name: '@types/d3-shape' },
  ],
});
