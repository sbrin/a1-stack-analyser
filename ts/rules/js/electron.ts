import { register } from '../../register.js';

register({
  tech: 'electron',
  name: 'Electron',
  type: 'tool',
  dependencies: [
    { type: 'npm', name: 'electron' },
    { type: 'githubAction', name: 'samuelmeuli/action-electron-builder' },
  ],
});
