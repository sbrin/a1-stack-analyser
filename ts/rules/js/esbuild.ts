import { register } from '../../register.js';

register({
  tech: 'esbuild',
  name: 'Esbuild',
  type: 'tool',
  dependencies: [{ type: 'npm', name: 'esbuild' }],
});
