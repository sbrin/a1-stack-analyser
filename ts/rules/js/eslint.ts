import { register } from '../../register.js';

register({
  tech: 'eslint',
  name: 'Eslint',
  type: 'tool',
  files: ['.eslintrc', '.eslintrc.cjs', '.eslintrc.json', '.eslintrc.js'],
  dependencies: [{ type: 'npm', name: 'eslint' }],
});
