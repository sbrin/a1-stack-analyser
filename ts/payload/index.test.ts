import { describe, expect, it, vi } from 'vitest';

import { matchAllFiles } from '../matchAllFiles.js';

import { Payload } from './index.js';

import '../autoload.js';

let id = 0;
vi.mock('../common/nid.ts', () => {
  return { nid: () => `${id++}` };
});

describe('Payload', () => {
  it('should detect nested dependencies', () => {
    const root = new Payload({ name: 'root', folderPath: '/' });
    root.addChild(
      new Payload({ name: 'PubSub', folderPath: '/', tech: 'gcp.pubsub' })
    );

    expect(root.toJson()).toMatchSnapshot();
  });

  it('should reuse correctly nested dependency', () => {
    const root = new Payload({ name: 'root', folderPath: '/' });
    root.addChild(
      new Payload({ name: 'GCP manual', folderPath: '', tech: 'gcp' })
    );
    root.addChild(
      new Payload({ name: 'PubSub', folderPath: '/', tech: 'gcp.pubsub' })
    );

    expect(root.childs[0].name).toBe('GCP manual');
    expect(root.childs[0].id).toBe('4');
    expect(root.childs[1].name).toBe('PubSub');
    expect(root.childs[1].id).toBe('5');
    expect(root.childs[1].inComponent!.id).toBe('4');
  });

  it('should dedup child and path', () => {
    const root = new Payload({ name: 'root', folderPath: '/' });
    root.addChild(
      new Payload({ name: 'GCP1', folderPath: 'foo', tech: 'gcp' })
    );
    root.addChild(
      new Payload({ name: 'GCP2', folderPath: 'foo', tech: 'gcp' })
    );
    root.addChild(
      new Payload({ name: 'GCP3', folderPath: 'bar.xml', tech: 'gcp' })
    );

    expect(root.childs).toHaveLength(1);
    expect(root.childs[0].path).toStrictEqual(new Set(['foo', 'bar.xml']));
    expect(root.toJson().childs[0].path).toStrictEqual(['foo', 'bar.xml']);
  });

  it('should dedup matched tech', () => {
    const root = new Payload({ name: 'root', folderPath: '/' });
    const res = matchAllFiles(
      [
        { name: 'index.ts', fp: 'index.ts', type: 'file' },
        { name: 'script.ts', fp: 'scripts.ts', type: 'file' },
      ],
      '/'
    );
    root.addTechs(res);

    expect(root.techs).toStrictEqual(new Set(['typescript']));
    expect(root.reason).toStrictEqual(new Set(['matched extension: .ts']));
  });
});
