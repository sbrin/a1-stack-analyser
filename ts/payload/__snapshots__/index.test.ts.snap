// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Payload > should detect nested dependencies 1`] = `
{
  "childs": [
    {
      "childs": [],
      "dependencies": [],
      "edges": [],
      "id": "2",
      "inComponent": null,
      "languages": {},
      "name": "Google Cloud Platform",
      "path": [
        "/",
      ],
      "reason": [
        "implicit: gcp.pubsub",
      ],
      "tech": "gcp",
      "techs": [],
    },
    {
      "childs": [],
      "dependencies": [],
      "edges": [],
      "id": "1",
      "inComponent": "2",
      "languages": {},
      "name": "PubSub",
      "path": [
        "/",
      ],
      "reason": [],
      "tech": "gcp.pubsub",
      "techs": [],
    },
  ],
  "dependencies": [],
  "edges": [],
  "id": "0",
  "inComponent": null,
  "languages": {},
  "name": "root",
  "path": [
    "/",
  ],
  "reason": [],
  "tech": null,
  "techs": [],
}
`;
